### PR TITLE
fix(vault): wire borrow/repay/redeem in Activity tab

### PR DESCRIPTION
**Pull request by jeremy-babylonlabs**
Activity tab previously rendered Deposit and Add Collateral as two rows for the same tx. Borrow, Repay, and Redeem did not appear at all because the indexer wasn't emitting them.

The indexer side is handled by babylon-vault-indexer #133 (borrow/repay/redeem emissions) and #141 (drops add_collateral / remove_collateral so the user-action projection has no paired duplicates). This PR is the corresponding frontend.

- Add Redeem to `ActivityType`; map indexer types `borrow`, `repay`, `redeem` to display labels.
- Drop `Add Collateral` / `Remove Collateral` from `ActivityType` and `GraphQLActivityType` — those rows are no longer emitted (#141), so the FE projection stays clean.
- Read `peginTxHash` from the existing `vaults` query rather than via a nested `vaultActivity → vault` relation (the relation isn't defined in `ponder.schema.ts`).
- Filter null `vaultId`s out of vault lookups so position-scoped rows (borrow/repay) don't break the join.
- Inject reserve metadata + borrow-app metadata via a deps interface so the service stays free of registry/network singletons.
- Format debt amounts via BigInt/string splitting to avoid `parseFloat` precision loss on high-decimals tokens; cap displayed fraction digits at 8.
- Degrade gracefully (`—` symbol, raw amount) when reserve metadata is missing, so a single malformed row never blanks the whole Activity tab.
- Parse `logIndex` from the indexer-generated id for deterministic same-tx ordering.

### CODE REVIEW

File: services/vault/src/services/activity/fetchActivities.ts [hunk 1]
```diff
@@ -109,40 +228,132 @@ export async function fetchUserActivities(
...
+        .filter((v): v is string => v != null),
+    ),
   );
+  // Resolve debt asset metadata with graceful degradation: if a reserve row
+  // is missing or its underlyingToken relation is null, fall back to the
+  // reserve's own decimals and an unknown-token symbol. One malformed row
+  // must never blank the whole Activity tab.
+  const reserveMap = new Map<string, { symbol: string; decimals: number }>();
+  if (reserveIds.length > 0) {
+    const reservesData = await graphqlClient.request<GraphQLReservesResponse>(
```
Comment (jrwbabylonlab) [inline services/vault/src/services/activity/fetchActivities.ts]: The comment above promises graceful degradation ("one malformed row must never blank the whole Activity tab"), but this call is an unconditional `await` — a failed `GET_RESERVES_BY_IDS` bubbles straight out of `fetchUserActivities`. So any user with at least one borrow/repay row loses the entire Activity tab (including all the vault-scoped rows that don't even need reserve metadata) whenever the reserve endpoint has an issue.

Suggested:
```ts
if (reserveIds.length > 0) {
  try {
    const reservesData = await graphqlClient.request<GraphQLReservesResponse>(GET_RESERVES_BY_IDS, { ids: reserveIds });
    for (const r of reservesData.aaveReserves.items) {
      reserveMap.set(r.id, { symbol: r.underlyingToken?.symbol ?? "—", decimals: r.underlyingToken?.decimals ?? r.decimals });
    }
  } catch (err) {
    logger.warn(`[activity] Reserve metadata fetch failed; borrow/repay rows will display raw amounts: ${err instanceof Error ? err.message : err}`);
  }
}
```
With the empty-reserveMap fallback already handled below (raw amount + `—` symbol), this delivers the degradation the comments promise.

One follow-up thought: when the fallback triggers, the raw `item.amount` is the smallest-unit integer (e.g. `1500000000` for 1.5 USDC). That will look alarming in the UI. Consider a safer display fallback — e.g. `formatAmount(item.amount, 18)` capped to some digits — so a transient reserve failure doesn't scare users with 10-digit amounts.

**Review by jrwbabylonlab — COMMENTED** [review]
Reviewed with codex.

**Review by gbarkhatov — COMMENTED** [review]
## Code Review

### Strengths

- **Well-scoped dedup logic** — groups by `(txHash, vaultId)`, only collapses known pairs. Multi-reserve borrow/repay survives because `vaultId` is null. Smart design.
- **Graceful degradation** — Missing reserve metadata falls back to raw amount + "—" symbol instead of crashing.
- **BigInt formatting** — `formatUnits` + BigInt string splitting instead of `parseFloat` is the right call for 18-decimal tokens.
- **Excellent test coverage** — 488 lines covering type mapping, dedup, null-safe fetching, precision, and degradation.

---

### Issues & Suggestions

#### 1. `parseLogIndex` is fragile — relies on undocumented id format

`fetchActivities.ts:43-48` — This is tightly coupled to the indexer's id schema (`${txHash}-${logIndex}-${type}-${vaultId}`). If the indexer ever changes the format, this silently returns `0` for all rows.

**Suggestion**: Add a comment with an example id, or request `logIndex` as a first-class field from the indexer in a follow-up.

#### 2. `dedupPairedActivities` has a subtle `typeToActivity` collision

`fetchActivities.ts:194`
```typescript
for (const a of bucket) typeToActivity.set(a.type, a);
```
If a bucket contains two rows with the same `type` (e.g., two `deposit` events in the same tx for the same vault — unlikely but theoretically possible from re-entrancy), the second overwrites the first. Probably fine in practice, but worth noting.

#### 3. BTC decimals hardcoded as `8`

`fetchActivities.ts:324`
```typescript
amountValue = formatAmount(item.amount, 8);
```
Per CLAUDE.md: "Extract all hardcoded numbers to named constants." Since `formatAmount` now accepts a `decimals` parameter, this is a good opportunity to extract it:
```typescript
const BTC_DECIMALS = 8;
// ...
amountValue = formatAmount(item.amount, BTC_DECIMALS);
```

#### 4. Hardcoded Aave assumption throws on missing metadata

`fetchActivities.ts:283-294` — The `throw` on missing Aave metadata is a potential runtime crash if `getApplication(AAVE_APP_ID)` returns `undefined` (e.g., feature flag disabling Aave). This is inconsistent with the PR's own resilience philosophy ("one bad row should never blank the whole tab").

**Suggestion**: Fall back to the `"Unknown App"` pattern used in the `else` branch rather than throwing.

#### 5. Sorting — `activities[idx]` assumes stable index mapping

`fetchActivities.ts:348`
```typescript
.map((row, idx) => ({ row, raw: activities[idx] }))
```
Correct because `rows` is built from `activities.map(...)`, but fragile — if the map is ever changed to `flatMap` or filtered, this breaks silently. Consider zipping `raw` + `row` during the initial map instead.

#### 6. Test mock dispatch based on string matching is brittle

`fetchActivities.test.ts:96-108` — Dispatching mocks via `String(query).includes("GetUserActivities")` works but is fragile if a query name is renamed or two queries share a substring. Not blocking.

#### 7. `parseLogIndex` defined far from its only usage

Defined at line 43, only used at line 356 in the sort comparator. Could be co-located with the sort logic for readability. Minor.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by gbarkhatov — COMMENTED** [review]
## PR Review

### CRITICAL: `GET_RESERVES_BY_IDS` failure kills the entire Activity tab

**File:** `fetchActivities.ts:267`

jrwbabylonlab already flagged this. The code comments promise graceful degradation ("one malformed row must never blank the whole Activity tab"), but `graphqlClient.request<GraphQLReservesResponse>(GET_RESERVES_BY_IDS, ...)` is an unguarded `await`. If the reserves query fails (network error, schema mismatch, timeout), the entire `fetchUserActivities` throws and the Activity tab goes blank for any user with at least one borrow/repay row.

**Fix:** Wrap the reserves fetch in try/catch and fall back to an empty `reserveMap`. The per-row fallback logic at line 319-322 already handles missing reserves gracefully — it just never gets a chance to run if the fetch itself throws.

```ts
if (reserveIds.length > 0) {
  try {
    const reservesData = await graphqlClient.request<GraphQLReservesResponse>(...);
    for (const r of reservesData.aaveReserves.items) {
      reserveMap.set(r.id, { ... });
    }
  } catch {
    // Degrade gracefully — per-row fallback will show raw amounts
  }
}
```

---

### MEDIUM: `typeToActivity` silently drops duplicate-type entries in the same bucket

**File:** `fetchActivities.ts:194`

If a bucket has two events of the same type (e.g., two `deposit` rows in the same tx+vault), `typeToActivity.set(a.type, a)` overwrites the first one. The dedup then only sees the last entry, and the dropped entry's `id` never enters `dropIds`, so it passes through unfiltered.

This is unlikely in practice (the indexer emits one event per type per tx+vault), but the code should either handle it explicitly (e.g., skip dedup for buckets with duplicate types) or add a comment explaining why duplicates can't happen.

---

### MEDIUM: Sequential awaits for independent fetches

**File:** `fetchActivities.ts:243-277`

`GET_VAULTS_BY_IDS` and `GET_RESERVES_BY_IDS` are independent queries issued sequentially. On a cold page load with many activities, this adds unnecessary latency. Use `Promise.all` (or `Promise.allSettled` if wrapping reserves in try/catch):

```ts
const [vaultsData, reservesData] = await Promise.all([
  vaultIds.length > 0 ? graphqlClient.request<GraphQLVaultsResponse>(...) : null,
  reserveIds.length > 0 ? graphqlClient.request<GraphQLReservesResponse>(...).catch(() => null) : null,
]);
```

---

### LOW: `parseLogIndex` is fragile — relies on undocumented id format

**File:** `fetchActivities.ts:43-48`

The function splits on `-` and takes `parts[1]` as the logIndex. This works today but is brittle — if the indexer id format ever changes, ordering silently degrades to `0` for all rows (due to the fallback). The comment documents the expected format, which is good. Consider logging a warning when the parse fails rather than silently defaulting to 0.

---

### LOW: BTC decimals hardcoded as `8`

**File:** `fetchActivities.ts:324`

```ts
amountValue = formatAmount(item.amount, 8);
```

Per CLAUDE.md's "No Magic Numbers" rule, this should be a named constant (e.g., `BTC_DECIMALS`).

---

### NITS

1. **Line 287-289:** `throw new Error("Aave application metadata not registered")` — this is a hard crash for a display function. If Aave config somehow isn't loaded, the entire activity tab dies. Consider falling back to a default app metadata object instead, consistent with the graceful degradation philosophy elsewhere in this PR.

2. **Line 159:** `BigInt(wholeRaw).toLocaleString("en-US")` — `BigInt.toLocaleString` behavior varies across engines. Verify this is tested in the target browser matrix.

---

### Positive observations

- **BigInt-safe formatting** — the `formatAmount` refactor correctly avoids `parseFloat` precision loss for high-decimals tokens. Well tested with the DAI 18-decimals case.
- **Dedup logic is well-scoped** — only collapses specific known pairs, leaves multi-reserve borrow/repay untouched.
- **Graceful per-row degradation** — when reserve metadata is missing for a single row, shows "—" and the raw amount rather than crashing.
- **Test coverage is thorough** — 488 lines covering type mapping, dedup pairs, null-safe fetching, formatting edge cases, and app metadata resolution.
- **`icon` field correctly omitted** for debt-denominated rows.

File: services/vault/src/services/activity/fetchActivities.ts [hunk 2]
```diff
@@ -109,40 +247,147 @@ export async function fetchUserActivities(
...
-    const applicationEntryPoint = vaultMap.get(item.vaultId);
-    const appMetadata = applicationEntryPoint
-      ? getApplicationMetadataByController(applicationEntryPoint)
-      : undefined;
+  // Fetch vault + reserve metadata in parallel. Reserve enrichment degrades
+  // gracefully on both outcomes: a network/schema failure leaves reserveMap
+  // empty (per-row fallback renders "—" + raw amount), and missing
+  // underlyingToken on an individual row falls back to the reserve's own
+  // decimals. One malformed row must never blank the whole Activity tab.
+  const [vaultsResult, reservesResult] = await Promise.all([
```
Comment (jrwbabylonlab) [inline services/vault/src/services/activity/fetchActivities.ts]: why 2 seperate graphql request when u can just do a single graphql call?

File: services/vault/src/services/activity/fetchActivities.ts [hunk 3]
```diff
@@ -109,40 +247,147 @@ export async function fetchUserActivities(
...
+            id: "unknown",
+            name: "Unknown App",
+            logoUrl: "/images/unknown-app.svg",
+          };
+    } else {
+      const applicationEntryPoint = item.vaultId
+        ? vaultMap.get(item.vaultId)
+        : undefined;
+      const appMetadata = applicationEntryPoint
+        ? getApplicationMetadataByController(applicationEntryPoint)
```
Comment (jrwbabylonlab) [inline services/vault/src/services/activity/fetchActivities.ts]: the application context layer should already fetched all static values, including this one.
I think you can pass in as an input into the method instead

File: services/vault/src/services/activity/fetchActivities.ts [hunk 4]
```diff
@@ -109,40 +247,147 @@ export async function fetchUserActivities(
...
+  const [vaultsResult, reservesResult] = await Promise.all([
+    vaultIds.length > 0
+      ? graphqlClient.request<GraphQLVaultsResponse>(GET_VAULTS_BY_IDS, {
+          vaultIds,
+        })
+      : Promise.resolve<GraphQLVaultsResponse | null>(null),
+    reserveIds.length > 0
+      ? graphqlClient
+          .request<GraphQLReservesResponse>(GET_RESERVES_BY_IDS, {
+            ids: reserveIds,
```
Comment (jrwbabylonlab) [inline services/vault/src/services/activity/fetchActivities.ts]: i think we only have a single reserve? can u double check?
I believe this is also fetched at the top layer contect provider. u can just pass in as method input instead

**Review by gbarkhatov — DISMISSED** [review]

File: services/vault/src/services/activity/fetchActivities.ts [hunk 5]
```diff
@@ -140,61 +165,196 @@ function resolveDisplayTx(item: GraphQLVaultActivityItem): {
...
+  );
+  const whole = BigInt(wholeRaw).toLocaleString("en-US");
+  const frac = fracRaw.slice(0, MAX_DISPLAY_FRACTION_DIGITS).replace(/0+$/, "");
+  return frac.length > 0 ? `${whole}.${frac}` : whole;
+}
+
+/**
+ * Collapse atomic tx-pairs emitted by the Aave adapter so a single user action
+ * doesn't surface as two Activity rows.
+ *
```
Comment (jrwbabylonlab) [inline services/vault/src/services/activity/fetchActivities.ts]: why is this the case? wouldn't the root cause on the indexer side instead? i.e why we emitting two same events?

Comment (jeremy-babylonlabs) [reply services/vault/src/services/activity/fetchActivities.ts]: i think the contract emits two events (PeginActivated and CollateralAdded), not sure if its a better idea to collapsing them on the indexer side

Comment (jrwbabylonlab) [reply services/vault/src/services/activity/fetchActivities.ts]: discussed in DM. this part of code smells. @jeremy-babylonlabs will verify with contract team and make a decision whether code change on the indexer side is needed

**Review by gbarkhatov — DISMISSED** [review]

**Review by jrwbabylonlab — APPROVED** [review]
